src/py21cmfast/drivers/param_config.py

```diff
@@ -172,63 +209,124 @@ def is_compatible_with(self, other: InputParameters) -> bool:
+                    f"InputParameters.from_output_struct got multiple values for {field}: {values}"
+                )
+            elif len(values) == 0:
+                if attrs.fields_dict(cls)[field].default:
+                    # If the parameter has a default, we want to remove them
+                    #    from the dict (after the loop)
+                    # TODO: this is messy, try to clean it up
+                    default_fields.append(field)
+                else:
+                    raise ValueError(
+                        f"InputParameters.from_output_struct got no values for required attribute {field}"
+                    )
+            # otherwise set value to the singleton
+            else:
+                input_params[field] = values.pop()
 
-        for inp in inputs[1:]:
-            this = this.merge(inp)
+        [input_params.pop(field) for field in default_fields]
+        return cls(**input_params)
 
-        return this
+    def check_output_compatibility(self, output_structs):
+        """Raises an error if the inputs are incompatible with the provided OutputStruct objects.
 
-    @classmethod
-    def from_output_structs(
-        cls, output_structs, redshift=None, **defaults
-    ) -> InputParameters:
-        """Generate a new InputParameters instance given a list of OutputStructs."""
-        input_params = []
-        for struct in output_structs:
-            if struct is not None:
-                ip_args = {
-                    k.lstrip("_"): getattr(struct, k, None)
-                    for k in struct._inputs
-                    if k.lstrip("_") in [field.name for field in attrs.fields(cls)]
-                }
-                input_params.append(cls(**ip_args))
-
-        if len(input_params) == 0:
-            return cls(**defaults)
-        else:
-            # Combine all the parameter structs from input boxes
-            out = cls.combine(input_params)
-            # Now combine with provided kwargs
-            return attrs.evolve(out.merge(cls(**defaults)), redshift=redshift)
+        Does not change the input struct.
+        """
+        InputParameters.from_output_structs(
+            output_structs,
+            node_redshifts=self.node_redshifts,
+            cosmo_params=self.cosmo_params,
+            user_params=self.user_params,
+            astro_params=self.astro_params,
+            flag_options=self.flag_options,
+            random_seed=self.random_seed,
+        )
+
+    def evolve_inputstructs(self, **kwargs):
+        """Return an altered clone of the `InputParamters` structs.
+
+        Unlike clone(), this function takes fields from the constituent `InputStruct` classes
+        and only overwrites those sub-fields instead of the entire field
+        """
+        cosmo_params = self.cosmo_params.clone(
+            **{
+                k: v
+                for k, v in kwargs.items()
+                if k in self.cosmo_params.asdict().keys()
```

Comment:
not too important, but I think doing the `.asdict()` inside the comprehension here might call the method for each iteration of the loop. You could save `cdict = self.cosmo_params.asdict()` outside the loop and use it inside the comprehension (and you don't need the `.keys()`)